### ISSUE TITLE
CentOS7 gem install mysql2 failed still using MariaDB 10.2

### ISSUE THREAD

**Issue by charlietag** [brianmario/mysql2#932]
Thanks for your hard work.
Due to failed as below, I have to use MariaDB 10.1 instead of 10.2

Mini environment as following ---> 
CentOS 7.4 - 64 bit
ruby 2.4.1
rails 5.1.4
MariaDB :
MariaDB-client-10.2.12-1.el7.centos.x86_64
MariaDB-devel-10.2.12-1.el7.centos.x86_64
MariaDB-compat-10.2.12-1.el7.centos.x86_64
MariaDB-server-10.2.12-1.el7.centos.x86_64
MariaDB-common-10.2.12-1.el7.centos.x86_64

<pre>
user$ rails new MyProject
...
Fetching mysql2 0.4.10
Installing mysql2 0.4.10 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /usr/local/lib/ruby/gems/2.4.0/gems/mysql2-0.4.10/ext/mysql2
/usr/local/bin/ruby -r ./siteconf20180107-57002-1jkcko8.rb extconf.rb
checking for rb_absint_size()... yes
checking for rb_absint_singlebit_p()... yes
checking for ruby/thread.h... yes
checking for rb_thread_call_without_gvl() in ruby/thread.h... yes
checking for rb_thread_blocking_region()... no
checking for rb_wait_for_single_fd()... yes
checking for rb_hash_dup()... yes
checking for rb_intern3()... yes
checking for rb_big_cmp()... yes
-----
Using mysql_config at /usr/bin/mysql_config
-----
checking for mysql.h... yes
checking for errmsg.h... yes
checking for SSL_MODE_DISABLED in mysql.h... no
checking for MYSQL_OPT_SSL_ENFORCE in mysql.h... yes
checking for MYSQL.net.vio in mysql.h... no
checking for MYSQL.net.pvio in mysql.h... yes
checking for MYSQL_ENABLE_CLEARTEXT_PLUGIN in mysql.h... yes
-----
Setting libpath to /usr/lib64/
-----
creating Makefile

current directory: /usr/local/lib/ruby/gems/2.4.0/gems/mysql2-0.4.10/ext/mysql2
make "DESTDIR=" clean

current directory: /usr/local/lib/ruby/gems/2.4.0/gems/mysql2-0.4.10/ext/mysql2
make "DESTDIR="
compiling client.c
In file included from ./mysql2_ext.h:35:0,
                 from client.c:1:
client.c: In function ‘rb_set_ssl_mode_option’:
./client.h:56:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   mysql_client_wrapper *wrapper; \
   ^
client.c:118:3: note: in expansion of macro ‘GET_CLIENT’
   GET_CLIENT(self);
   ^
client.c:119:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   int val = NUM2INT( setting );
   ^
client.c:145:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
client.c: At top level:
cc1: warning: unrecognized command line option "-Wno-self-assign" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-constant-logical-operand" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-parentheses-equality" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-tautological-compare" [enabled by default]
compiling infile.c
compiling mysql2_ext.c
compiling result.c
result.c: In function ‘rb_mysql_result_alloc_result_buffers’:
result.c:326:40: warning: assignment from incompatible pointer type [enabled by default]
     wrapper->result_buffers[i].is_null = &wrapper->is_null[i];
                                        ^
result.c:328:40: warning: assignment from incompatible pointer type [enabled by default]
     wrapper->result_buffers[i].error   = &wrapper->error[i];
                                        ^
result.c: At top level:
cc1: warning: unrecognized command line option "-Wno-self-assign" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-constant-logical-operand" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-parentheses-equality" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-tautological-compare" [enabled by default]
compiling statement.c
statement.c: In function ‘rb_raise_mysql2_stmt_error’:
statement.c:50:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   VALUE rb_error_msg = rb_str_new2(mysql_stmt_error(stmt_wrapper->stmt));
   ^
statement.c:57:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   rb_encoding *default_internal_enc = rb_default_internal_encoding();
   ^
In file included from ./mysql2_ext.h:35:0,
                 from statement.c:1:
statement.c: In function ‘execute’:
./client.h:56:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   mysql_client_wrapper *wrapper; \
   ^
statement.c:273:3: note: in expansion of macro ‘GET_CLIENT’
   GET_CLIENT(stmt_wrapper->client);
   ^
statement.c:405:13: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
             VALUE rb_val_as_string = rb_funcall(argv[i], intern_to_s, 0);
             ^
In file included from ./mysql2_ext.h:35:0,
                 from statement.c:1:
statement.c: In function ‘fields’:
./client.h:56:3: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
   mysql_client_wrapper *wrapper; \
   ^
statement.c:481:3: note: in expansion of macro ‘GET_CLIENT’
   GET_CLIENT(stmt_wrapper->client);
   ^
statement.c: At top level:
cc1: warning: unrecognized command line option "-Wno-self-assign" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-constant-logical-operand" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-parentheses-equality" [enabled by default]
cc1: warning: unrecognized command line option "-Wno-tautological-compare" [enabled by default]
linking shared-object mysql2/mysql2.so
/usr/bin/ld: cannot find -lmariadb
collect2: error: ld returned 1 exit status
make: *** [mysql2.so] Error 1
  
make failed, exit code 2
  
Gem files will remain installed in /usr/local/lib/ruby/gems/2.4.0/gems/mysql2-0.4.10 for inspection.
Results logged to /usr/local/lib/ruby/gems/2.4.0/extensions/x86_64-linux/2.4.0-static/mysql2-0.4.10/gem_make.out
  
An error occurred while installing mysql2 (0.4.10), and Bundler cannot continue.
Make sure that `gem install mysql2 -v '0.4.10'` succeeds before bundling.
  
In Gemfile:
  mysql2
         run  bundle exec spring binstub --all
bundler: command not found: spring
Install missing gem executables with `bundle install`
</pre>

**Comment by sodabrew** [comment]
The latest MariaDB 10.2 fixes have not been released yet. Look for 0.4.11 and 0.5.0 coming in the next 1-2 months!

**Comment by charlietag** [comment]
Thank you so much !
I'm expecting 0.5.0.
By the way, In readme, I found MariaDB 10.2 is tested 
<---- Does this mean that it was tested in Ubuntu ? So that my CentOS cannot work.
<pre>
This gem is tested with the following MySQL and MariaDB versions:

MySQL 5.5, 5.6, 5.7, 8.0
MySQL Connector/C 6.0 and 6.1 (primarily on Windows)
MariaDB 5.5, 10.0, 10.1, 10.2
</pre>

**Comment by sodabrew** [comment]
It'll work fine on CentOS. Just blocked on me sitting down with time to cut the release and sort out the Windows builds.

**Comment by luisbrandao** [comment]
I am suck with the same:

linking shared-object mysql2/mysql2.so
/bin/ld: cannot find -lmariadb
collect2: error: ld returned 1 exit status
make: ** [mysql2.so] Erro 1

make failed, exit code 2

How can i  workaround it?

**Comment by sodabrew** [comment]
The workarounds are waiting until I get my next time window for working on the mysql2 release or downgrading to MariaDB 10.1. My apologies for the delay!

**Comment by charlietag** [comment]
@sodabrew 
Sorry, not pushing this.
Just a quick ask. 
May I know when will it be ready for MariaDB 10.2 ?
Or it would be better to use MariaDB 10.1 ?

Thanks for your big efforts again.

**Comment by sodabrew** [comment]
Just posted mysql2 0.5.0 with improved support for MariaDB 10.2 and 10.3.

**Comment by charlietag** [comment]
@sodabrew 
Thanks so much for your big effort !  I'm testing it!

**Comment by UAnton** [comment]
mysql2 0.5.0 not working with 10.2 and 10.3

**Comment by sodabrew** [comment]
Please provide more information. MariaDB 10.0, 10.1, 10.2, and 10.3 are all in the CI matrix for this gem, so I am confident that they do work. If you are experiencing a problem, please include details about that problem.

**Comment by sodabrew** [comment]
@luisbrandao In another ticket, someone had posted that you must install the `mariadb-shared` package to get the `libmariadb.so` library.

**Comment by mdzidic** [comment]
Hello,

I have similar problem installing mysql2-0.5.1

`ruby 2.4.1p111 (2017-03-22 revision 58053) [x86_64-linux]`
`mysql  Ver 15.1 Distrib 10.2.15-MariaDB, for Linux (x86_64) using readline 5.1`

- Package MariaDB-devel-10.2.15-1.el7.centos.x86_64 already installed and latest version
- Package MariaDB-compat-10.2.15-1.el7.centos.x86_64 already installed and latest version


Error log:
```Building native extensions.  This could take a while...
ERROR:  Error installing mysql2:
	ERROR: Failed to build gem native extension.

    current directory: /opt/sensu/embedded/lib/ruby/gems/2.4.0/gems/mysql2-0.5.1/ext/mysql2
/opt/sensu/embedded/bin/ruby -r ./siteconf20180618-24143-1iiqmim.rb extconf.rb
checking for rb_absint_size()... yes
checking for rb_absint_singlebit_p()... yes
checking for rb_wait_for_single_fd()... yes
-----
Using mysql_config at /usr/bin/mysql_config
-----
checking for mysql.h... yes
checking for errmsg.h... yes
checking for SSL_MODE_DISABLED in mysql.h... no
checking for MYSQL_OPT_SSL_ENFORCE in mysql.h... yes
checking for MYSQL.net.vio in mysql.h... no
checking for MYSQL.net.pvio in mysql.h... yes
checking for MYSQL_ENABLE_CLEARTEXT_PLUGIN in mysql.h... yes
checking for SERVER_QUERY_NO_GOOD_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_NO_INDEX_USED in mysql.h... yes
checking for SERVER_QUERY_WAS_SLOW in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_ON in mysql.h... yes
checking for MYSQL_OPTION_MULTI_STATEMENTS_OFF in mysql.h... yes
checking for my_bool in mysql.h... yes
-----
Setting libpath to /usr/lib64/
-----
creating Makefile

current directory: /opt/sensu/embedded/lib/ruby/gems/2.4.0/gems/mysql2-0.5.1/ext/mysql2
make "DESTDIR=" clean

current directory: /opt/sensu/embedded/lib/ruby/gems/2.4.0/gems/mysql2-0.5.1/ext/mysql2
make "DESTDIR="
compiling client.c
compiling infile.c
compiling mysql2_ext.c
compiling result.c
compiling statement.c
linking shared-object mysql2/mysql2.so
/usr/bin/ld: cannot find -lmariadb
collect2: error: ld returned 1 exit status
make: *** [mysql2.so] Error 1

make failed, exit code 2

Gem files will remain installed in /opt/sensu/embedded/lib/ruby/gems/2.4.0/gems/mysql2-0.5.1 for inspection.
Results logged to /opt/sensu/embedded/lib/ruby/gems/2.4.0/extensions/x86_64-linux/2.4.0/mysql2-0.5.1/gem_make.out
```

**Comment by charlietag** [comment]
@mdzidic 
Try ``yum install MariaDB-shared``
#949

**Comment by mdzidic** [comment]
@charlietag package `MariaDB-shared` fixed this issue, thanks!

**Comment by charlietag** [comment]
@mdzidic MariaDB 10.3 is just tested. It also works for MariaDB 10.3 on CentOS 7.4.
By the way, since you're also using CentOS7. 
Maybe you can try my [os_preparation](https://github.com/charlietag/os_preparation)
It's hard to find CentOS fellows nowadays in rails world.  Good to hear you're using CentOS 7.

**Comment by Yu-Chieh-Henry-Yang** [comment]
Hello @charlietag , and thank you for the solution. It solves my issue. Can I ask how would a person know `yum install MariaDB-shared` will solve the problem? And do you know why it solves the problem?

**Comment by glouton** [comment]
Hi there,
Months later installing MariaDB-shared as per @charlietag suggestion works for me with MariaDB 10.4.11 on CentOS 7.7
Cheers!
